src/edgeit.jl

```diff
@@ -0,0 +1,66 @@
+import Base: start, next, done, length, show,
+            in, ==
+
+type EdgeItState
+    it::Int
+    v::Int
+    k::Int
+end
+
+type EdgeIt
+    m::Int
+    adj::Vector{Vector{Int}}
+    start::EdgeItState # =[it, v, k]
+    directed::Bool
+end
+
+edge_it(g::Graph) = EdgeIt(ne(g), g.fadjlist, EdgeItState(1,1,1), false)
+edge_it(g::DiGraph) = EdgeIt(ne(g), g.fadjlist, EdgeItState(1,1,1), true)
+
+start(eit::EdgeIt) = eit.start
+done(eit::EdgeIt, state::EdgeItState) = state.it > eit.m
+function next(eit::EdgeIt, state::EdgeItState)
+    # println("$eit, $state")
+    assert(state.v <= length(eit.adj))
+    u = 0
+    while state.k <= length(eit.adj[state.v])
+        u  = eit.adj[state.v][state.k]
+        eit.directed && break
+        u >= state.v && break
+        state.k += 1
+    end
+    while state.k > length(eit.adj[state.v])
+        state.k = 1
+        state.v += 1
+        while state.k <= length(eit.adj[state.v])
+            u  = eit.adj[state.v][state.k]
+            eit.directed && break
+            u >= state.v && break
+            state.k += 1
+        end
+    end
+    e = !eit.directed && u < state.v ? Edge(u, state.v) : Edge(state.v, u)
+    state.k += 1
+    state.it += 1
+    return (e, state)
+end
+length(eit::EdgeIt) = eit.m
+
+function in(e::Edge, eit::EdgeIt)
+    s = src(e)
+    t = dst(e)
+    n = length(eit.adj)
+    !(1 <=  s <= n) && return false
+    !(1 <=  t <= n) && return false
+    return t in eit.adj[s]
```

Comment:
You're quite right about the tradeoffs. It's one of the reasons there are so many different ways to represent graphs. :)

Most of my work is in sparse graphs as well, so it won't affect ME too much, but I don't want it to be a surprise for anyone else. If you find calls to `has_edge()` that's probably where you'll get tripped up.